typed-racket-test/succeed/test-struct-info.rkt

```diff
@@ -0,0 +1,39 @@
+#lang racket/base
+(module test racket/base
+  (require (for-syntax racket/base syntax/parse racket/struct-info racket/match rackunit))
+
+  (define-syntax (check-constr stx)
+    (syntax-parse stx
+      [(_ sname:id expected-constr:id)
+       (define si (extract-struct-info (syntax-local-value #'sname (lambda () #f))))
+       (match-define (list* struct-desc constr rst) si)
+       (check-equal? (syntax-e constr) (syntax-e #'expected-constr))
+       #'(begin)]))
+  (provide check-constr))
+
+(module typed typed/racket/base
+  (require (submod ".." test))
+  (struct kiwi ())
+  (struct apple () #:constructor-name make-apple)
+  (struct pear () #:type-name PEAR)
+  (struct dragon-fruit () #:type-name DragonFruit #:constructor-name make-dragon-fruit)
+  (provide (struct-out apple)
+           (struct-out kiwi)
+           (struct-out PEAR)
+           (struct-out DragonFruit))
+
+  (check-constr apple make-apple)
+  (check-constr kiwi kiwi)
+  (check-constr apple make-apple)
+  (check-constr pear pear)
+  (check-constr dragon-fruit make-dragon-fruit)
+  (check-constr DragonFruit make-dragon-fruit)
+  (check-constr PEAR pear))
+
+(require 'typed)
+(require 'test)
+(check-constr apple make-apple)
```

Comment:
I feel like this test could be simpler/more direct. For example, is `(object-name apple)` always `'make-apple` here? Or can we just check that everything works correctly as a procedure?

Reply:
Do you mean we use `object-name` at phase 0? my understanding is 
1. if v is simply a structure type identifier and not a constructor name, then we will see racket/typed racket complains identifier for static struct-type information cannot be used as an expression before `(object-name v)` run
2. `object-name` doesn't return the constructor name. E.g `(object-name (make-apple))` will return `'apple`
3. If I understand you correctly, everything already almost worked correctly as a procedure before this patch. Previously we used `export-id` as the constructor name. In this case, the struct info of `(struct apple () #:constructor-name make-apple)`  incorrectly had some form of #'apple as the constructor. But we found no errors from running TR's tests, buiding packages and running their tests until `delay-pure` which directly used the struct info of typed structure types.

Reply:
I thinking of this code:

```
#lang racket

(module test racket/base
  (require  (for-syntax racket/base syntax/parse racket/struct-info racket/match))

  (define-syntax (check-constr stx)
    (syntax-parse stx
      [(_ sname:id)
       (define si (extract-struct-info (syntax-local-value #'sname (lambda () #f))))
       (match-define (list* struct-desc constr rst) si)
       constr]))
  (provide check-constr))
(module t typed/racket/base
  (require (submod ".." test) typed/rackunit)
  (struct kiwi ())
  (struct apple () #:constructor-name make-apple)
  (struct pear () #:type-name PEAR)
  (struct dragon-fruit () #:type-name DragonFruit #:constructor-name make-dragon-fruit)

  (check-equal? (check-constr apple) 'make-apple)

  (provide (struct-out apple)
           (struct-out kiwi)
           (struct-out PEAR)
           (struct-out DragonFruit)))

(require 'test 't rackunit)
(check-equal? (check-constr apple) 'make-apple)
```

Reply:
Ah, this is better